src/libYARP_OS/include/yarp/os/PeriodicThread.h

```diff
@@ -0,0 +1,223 @@
+/*
+ * Copyright (C) 2006-2018 Istituto Italiano di Tecnologia (IIT)
+ * Copyright (C) 2006-2010 RobotCub Consortium
+ * All rights reserved.
+ *
+ * This software may be modified and distributed under the terms of the
+ * BSD-3-Clause license. See the accompanying LICENSE file for details.
+ */
+
+#ifndef YARP_OS_PERIODICTHREAD_H
+#define YARP_OS_PERIODICTHREAD_H
+
+#include <yarp/os/Runnable.h>
+
+/**
+ * \ingroup key_class
+ *
+ * An abstraction for a periodic thread.
+ */
+namespace yarp {
+namespace os {
+
+class YARP_OS_API PeriodicThread
+{
+public:
+
+    /**
+     * Constructor.  Thread begins in a dormant state.  Call PeriodicThread::start
+     * to get things going.
+     * @param period The period in seconds [sec] between
+     * successive calls to the RateThread::run method
+     * (remember you need to call RateThread::start first
+     * before anything happens)
+     * @param useSystemClock whether the thread should always
+     * use the system clock, or depend on the current
+     * configuration of the network.
+     */
+    explicit PeriodicThread(double period, bool useSystemClock = false);
```

Comment:
Not a big problem, but see https://abseil.io/tips/94 about bool parameters.

Reply:
Are you suggesting to use an enum here? I think I :heart: the idea:
```c++
enum class ShouldUseSystemClock { No = 0, Yes = 1 };
explicit PeriodicThread(double period, ShouldUseSystemClock useSystemClock = ShouldUseSystemClock::No);
```

And then when you use it

```c++
PeriodicThread p(0.01, ShouldUseSystemClock:Yes)
```

Reply:
I really like the idea, but I thought that this would be problematic for SWIG. However, it seems that nested classes are now relativly supported in SWIG 3: http://www.swig.org/Doc3.0/SWIGPlus.html#SWIGPlus_nested_classes , but clearly that should be tested before giving it for granted.

Reply:
@traversaro Do you think you have time to test SWIG?

Reply:
No, if you want to use this enum solution please go ahead and we will eventually fix it for SWIG.

Reply:
I like the idea too.
If we want to use enums, there is already the `yarpClockType` defined in Time.h, maybe we can use that one

Reply:
The problem with `yarpClockType` is that has a different meaning... it is necessary to choose the clock to use, and it can be network, system or custom and it must be passed in the init phase.

Here we are choosing whether to use the system clock, or to use the clock set up during init(), i.e. it should be a "bool", a third option is confusing.

I think it might be a good idea to have this new enum in `Time.h` though, it might be useful for other stuff (i.e. `Timer`, `RFModule`, ...)